### PR TITLE
CLAModel uses Anomaly region

### LINKED ISSUE #2073: Improve AnomalyRegion and integration with CLAModel
this is a follow-up on https://github.com/numenta/nupic/issues/2058
The work to be done for AnomalyRegion: 
* the region creates an instance of Anomaly
* allows to set all parameters
* integrates to CLAModel for anomaly computation
* integrates to CLAModel for anomaly params swarming
* serialization

### PR DESCRIPTION
**Pull request by breznak**
Fixes #2073 

TODO:
- [x] CLAModel uses AnomalyRegion
- [x] pass `anomalyParams` from CLAModel to AnomalyRegion
- [x] pass rawInput 
- [ ] pass timestamps if possible 
- [ ] (opt) remove anomalyScore fields from TP/SP

### CODE REVIEW

File: nupic/frameworks/opf/clamodel.py [hunk 1]
```diff
@@ -1135,6 +1125,33 @@ def __createCLANetwork(self, sensorParams, spEnable, spParams, tpEnable,
       )
       self._addAnomalyClassifierRegion(n, anomalyClParams, spEnable, tpEnable)
 
+    # add Anomaly region
+    if anomalyParams and tpEnable:
+      # ignore these params, used for AnomalyClassifierRegion, but shared under anomalyParams{}
+      args = copy.deepcopy(anomalyParams)
+      args.pop("anomalyCacheRecords")
```
Comment (breznak) [inline nupic/frameworks/opf/clamodel.py]: I have to get rid of these, as AnomalyRegion complained for "unknown param"

File: nupic/frameworks/opf/clamodel.py [hunk 2]
```diff
@@ -1135,6 +1125,33 @@ def __createCLANetwork(self, sensorParams, spEnable, spParams, tpEnable,
...
+      args.pop("autoDetectThreshold")
+      args.pop("autoDetectWaitRecords")
+
+      self.__logger.debug("Adding AnomalyRegion; params: %r" % args)
+      n.addRegion("Anomaly", "py.AnomalyRegion", json.dumps({})) #FIXME how pass args as kwargs here?
```
Comment (breznak) [inline nupic/frameworks/opf/clamodel.py]: I'd like to pass anomalyParams as kwargs, how to do that?

File: nupic/frameworks/opf/clamodel.py [hunk 3]
```diff
@@ -1135,6 +1125,33 @@ def __createCLANetwork(self, sensorParams, spEnable, spParams, tpEnable,
...
+
+      # link SP (or sensor) output as activeColumns
+      if spEnable:
+        n.link("SP", "Anomaly", "UniformLink", "",
+                 srcOutput="bottomUpOut", destInput="activeColumns")
+      else:
+        activeColumns = sensor.getOutputData('dataOut').nonzero()[0]
+        n.link("sensor", "Anomaly", "UniformLink", "",
+               srcOutput="dataOut", destInput="activeColumns")
```
Comment (breznak) [inline nupic/frameworks/opf/clamodel.py]: get rid of this (SP|Sensor) when #2081 is done

File: nupic/regions/TPRegion.py [hunk 4]
```diff
@@ -565,6 +565,7 @@ def getBaseSpec(cls):
           regionLevel=True,
           isDefaultOutput=False),
 
+        # TODO remove anomalyScore from TPRegion, it can be provided by AnomalyRegion if needed
         anomalyScore = dict(
```
Comment (breznak) [inline nupic/regions/TPRegion.py]: same for SPRegion

**Comment by breznak** [comment]
@oxtopus @scottpurdy please review

**Comment by david-ragazzi** [comment]
@breznak 

`TypeError: unsupported operand type(s) for +=: 'float' and 'NoneType'`

File: nupic/frameworks/opf/clamodel.py (hunk 2 above)
Comment (scottpurdy) [inline nupic/frameworks/opf/clamodel.py]: put them in the dict that is serialized to json

**Comment by scottpurdy** [comment]
@breznak - This will lose some of the anomaly-related functionality you have added recently. So maybe you will want to add an anomaly region that takes other inputs and supports all of the different anomaly modes and use that region instead?

**Comment by rhyolight** [comment]
> Can this PR be labelled as "ready" after Travis is green?

There are undone `TODO` items in the PR description, so I assumed it was still a WIP.

**Comment by breznak** [comment]
yes..still WIP. 

@scottpurdy 
> This will lose some of the anomaly-related functionality you have added recently. So maybe you will want to add an anomaly region that takes other inputs and supports all of the different anomaly modes and use that region instead?

The normal input are the 2 bit arrays, the only extra input can be: `rawInput` (+timestamp) for `likelihood`; 
I will need to: 
- [ ] just add (=copy) params from Anomaly class to AnomalyRegion
- [ ] add the (opt.) `rawInput` input for likelihood 

to support the current functionality. I don't know how you mean in with the new region(?), the problem I see with specific regions is again in swarming..

File: nupic/regions/AnomalyRegion.py [hunk 5]
```diff
@@ -57,6 +57,15 @@ def getSpec(cls):
                 "isDefaultInput": False,
                 "requireSplitterMap": False,
             },
+            "rawInput": {
+                "description": "The raw values coming from the input sensor.",
+                "regionLevel": True,
+                "dataType": "Real32", #TODO what type to set here? can be anything
```
Comment (breznak) [inline nupic/regions/AnomalyRegion.py]: please advise here

File: nupic/regions/AnomalyRegion.py [hunk 6]
```diff
@@ -84,9 +100,12 @@ def initialize(self, inputs, outputs):
 
   def compute(self, inputs, outputs):
     activeColumns = inputs["activeColumns"].nonzero()[0]
+    rawInput = inputs.get("rawInput", None)
+    timestamp = inputs.get("timestamp", None)
+    print "RAW=", rawInput
```
Comment (breznak) [inline nupic/regions/AnomalyRegion.py]: parse rawInput & timestamps correctly from the inputs

**Comment by breznak** [comment]
Blocked by: #2266

**Comment by scottpurdy** [comment]
@breznak - can we close this for now? We first need to make changes to anomaly region. Then we should give it some breathing room before migrating the CLAModel.

**Comment by scottpurdy** [comment]
@breznak - Thanks, once we get the anomaly region supporting the different types of anomalies then we can make the switch for the CLAModel

**Pull request closed without merging**